nengo/solvers.py

```diff
@@ -116,7 +116,8 @@ class _LstsqNoiseSolver(Solver):
     solver = LeastSquaresSolverParam("solver")
 
     def __init__(self, weights=False, noise=0.1, solver=lstsq.Cholesky()):
-        """
+        """Create new solver.
```

Comment:
I think this (and the other changes to `__init__` methods) will result in some pretty awkward documentation in the end ... Sphinx concatenates together the class docstring and the `__init__` docstring, so this sentence will appear in after the class docstring because we set `autoclass_content = "both"` in `conf.py`.

At the same time, I agree that if anyone happened to look at the `__init__` docstring in isolation, it would be awkward to just see the parameter lists and whatnot.

The reason that we don't just put the parameters etc in the class docstring is because there's a base class implementation that many (but not all) solvers use. So some solvers override the basic description (class docstring) but not the parameter / attribute lists (`__init__` docstring).

So what we have right now basically results in good Sphinx docs, but is not great for reading the docstrings in source or when using `help`.

One possible way forward that still lets us reduce duplication of parameter lists would be to use a simple decorator, as we did in pytest-allclose (see https://github.com/nengo/pytest-allclose/blob/master/pytest_allclose/plugin.py#L9). That way everything will be in the class docstring, and we simply leave `__init__` undocumented (if I wanted to know about a class, I definitely wouldn't think to do `help(ClassName.__init__)`.

Any other ideas?

Reply:
The simple decorator seems straightforward, even though it's a bit of work to set up.

The only other thing I can think of is a fancy decorator that goes through base classes, finds the "Parameters" (or other sections) in the docstring, and adds the things in there to the parameter list. Seems like something Sphinx or some other package ought to have. But I don't think it's worth us putting in that work at this point.

Reply:
Implemented the somewhat simple decorators; the Nnls one was a bit less simple, but reduced any duplication, which is better than what is currently done.